ee/session_recordings/session_summary/output_data.py

```diff
@@ -0,0 +1,463 @@
+        json_content: dict = yaml.safe_load(raw_content.strip("```yaml\n").strip("```").strip())  # noqa: B005
+    except Exception as e:
+        raise SummaryValidationError(
+            f"Error loading YAML content into JSON when summarizing session_id {session_id}: {e}"
+        )
+    # Validate the LLM output against the schema
+    raw_session_summary = RawSessionSummarySerializer(data=json_content)
+    if not raw_session_summary.is_valid():
+        raise SummaryValidationError(
+            f"Error validating LLM output against the schema when summarizing session_id {session_id}: {raw_session_summary.errors}"
+        )
+    segments = raw_session_summary.data.get("segments")
+    if not segments:
+        # If segments aren't generated yet - return the current state
+        return raw_session_summary
+    segments_indices = [segment.get("index") for segment in segments]
+    key_actions = raw_session_summary.data.get("key_actions")
+    if not key_actions:
+        # If key actions aren't generated yet - return the current state
+        return raw_session_summary
+    for key_action_group in key_actions:
+        key_group_segment_index = key_action_group.get("segment_index")
+        if key_group_segment_index is None:
+            # If key group segment index isn't generated yet - skip this group
+            continue
+        # Ensure that LLM didn't hallucinate segments
+        if key_group_segment_index not in segments_indices:
+            raise ValueError(
+                f"LLM hallucinated segment index {key_group_segment_index} when summarizing session_id {session_id}: {raw_session_summary.data}"
+            )
+        key_group_events = key_action_group.get("events")
+        if not key_group_events:
+            # If key group events aren't generated yet - skip this group
+            continue
+        for event in key_group_events:
+            # Ensure that LLM didn't hallucinate events
+            event_id = event.get("event_id")
+            if not event_id or len(event_id) != 8:
+                # If event ID isn't fully generated yet - skip this event
+                continue
+            # TODO: Allow skipping some events (even if not too many to speed up the process
+            if event_id not in allowed_event_ids:
+                raise ValueError(
+                    f"LLM hallucinated event_id {event_id} when summarizing session_id "
+                    f"{session_id}: {raw_session_summary.data}"
+                )
+    return raw_session_summary
+
+
+# TODO Rework the logic, so events before the recording are marked as "LOAD", not 00:00
+def calculate_time_since_start(session_timestamp: str, session_start_time: datetime | None) -> int | None:
+    if not session_start_time or not session_timestamp:
+        return None
+    timestamp_datetime = datetime.fromisoformat(session_timestamp)
+    # TODO Check why the event could happen before the session started
```

Comment:
assuming this is a "how could this happen" thing

(otherwise ignore this context 😊)

we can load the SDK in the browser and send events before rrweb has finished lazy loading so it's totally normal to have events before the start

and we prioritise the (smaller) analytics events when flushing on page close so it's normal to have events after session recording end

that's not counting folk with triggers set to purposefully start recording late, or who add session id to backend events which don't _have_ to happen at the same time as recording

Reply:
Yup, thank you. I have TODO to keep negative time and convert them to "LOAD" in the player instead, but it's a follow-up, as I want to land this PR as soon as possible and switch to smaller iterations that are easier to review 😄 

Will remove the comment.

Reply:
[`427a71f` (#31394)](https://github.com/PostHog/posthog/pull/31394/commits/427a71f91700ecf063821ef645678af08f6c19c6)